sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/FluxInputStream.java

```diff
@@ -0,0 +1,242 @@
+    private final Lock lock;
+    private final Condition dataAvailable;
+
+    private IOException lastError;
+
+    /**
+     * Creates a new FluxInputStream
+     *
+     * @param data The data to subscribe to and read from.
+     */
+    public FluxInputStream(Flux<ByteBuffer> data) {
+        this.subscribed = false;
+        this.fluxComplete = false;
+        this.waitingForData = false;
+        this.data = data;
+        this.lock = new ReentrantLock();
+        this.dataAvailable = lock.newCondition();
+    }
+
+    @Override
+    public int read() throws IOException {
+        byte[] ret = new byte[1];
+        int count = read(ret, 0, 1);
+        return count == -1 ? -1 : ret[0];
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        validateParameters(b, off, len);
+
+        /* If len is 0, then no bytes are read and 0 is returned. */
+        if (len == 0) {
+            return 0;
+        }
+        /* Attempt to read at least one byte. If no byte is available because the stream is at end of file,
+           the value -1 is returned; otherwise, at least one byte is read and stored into b. */
+
+        /* Not subscribed? subscribe and block for data */
+        if (!subscribed) {
+            blockForData();
+        }
+        /* Now, we have subscribed. */
+        /* At this point, buffer should not be null. If it is, that indicates either an error or completion event
+           was emitted by the Flux. */
+        if (this.buffer == null) { // Only executed on first subscription.
+            if (this.lastError != null) {
+                throw logger.logThrowableAsError(this.lastError);
+            }
+            if (this.fluxComplete) {
+                return -1;
+            }
+            this.buffer = new ByteArrayInputStream(new byte[0]);
+        }
+
+        /* Now we are guaranteed that buffer is SOMETHING. */
+        /* No data is available in the buffer.  */
+        if (this.buffer.available() == 0) {
+            /* If the flux completed, there is no more data available to be read from the stream. Return -1. */
+            if (this.fluxComplete) {
+                return -1;
+            }
+            /* Block current thread until data is available. */
+            blockForData();
+        }
+
+        /* Data available in buffer, read the buffer. */
+        if (this.buffer.available() > 0) {
+            return this.buffer.read(b, off, len);
+        }
+
+        /* If the flux completed, there is no more data available to be read from the stream. Return -1. */
+        if (this.fluxComplete) {
+            return -1;
+        } else {
+            throw logger.logExceptionAsError(new IllegalStateException("An unexpected error occurred. No data was "
+                + "read from the stream but the stream did not indicate completion."));
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        subscription.cancel();
```

Comment:
Is there any value in checking if the flux has completed before cancelling the subscription? Or is this a no-op in that case?

Reply:
I don't think it would matter if the flux completed or not because the user is done with the stream regardless

Reply:
Yea true. I more was just worried about the flux throwing an exception if you cancel a completed subscription.

Reply:
I don;t see any documentation about it throwing if the subscription completed. https://www.reactive-streams.org/reactive-streams-1.0.0-javadoc/org/reactivestreams/Subscription.html#cancel--